File: 07-langs_particularities/3-scala/src/main/scala/codely/contexts/rrss/users/application/find/UserFinder.scala
package codely.contexts.rrss.users.application.find

import codely.contexts.rrss.users.domain.{User, UserDoesNotExistError, UserId, UserRepository}

class UserFinder(repository: UserRepository) {
	def find(id: String): Either[UserDoesNotExistError, User] = {
		repository.search(UserId(id)) match {
			case None => Left(UserDoesNotExistError(id))
			case Some(user) => Right(user)
		}
	}
}

File: 07-langs_particularities/2-java/src/main/java/tv/codely/contexts/rrss/users/application/UserFinder.java
package tv.codely.contexts.rrss.users.application;

import tv.codely.contexts.rrss.users.domain.User;
import tv.codely.contexts.rrss.users.domain.UserDoesNotExistError;
import tv.codely.contexts.rrss.users.domain.UserRepository;

public record UserFinder(UserRepository repository) {
    public User find(String id) throws UserDoesNotExistError {
        var user = repository.search(id);

        if (user == null) {
            throw new UserDoesNotExistError(id);
        }

        return user;
    }
}
